### PR TITLE
Implement direct connection codes

### PR DESCRIPTION
**Pull request by vabold**
Thank you @stblr for doing all of the hard parts already.

### CODE REVIEW

File: payload/game/ui/DirectConnectionPage.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#include "DirectConnectionPage.hh"
+
+#include "game/ui/SectionManager.hh"
```
Comment (stblr) [inline payload/game/ui/DirectConnectionPage.cc:3]: One blank line after this.

File: payload/sp/cs/RoomClient.cc [hunk 2]
```diff
@@ -95,9 +95,11 @@ RoomClient *RoomClient::Instance() {
     return s_instance;
 }
 
-RoomClient::RoomClient(u32 localPlayerCount)
+RoomClient::RoomClient(u32 localPlayerCount, u32 ip, u16 port, u16 passcode)
     : m_localPlayerCount(localPlayerCount), m_state(State::Connect),
-      m_socket(0x7f000001, 21330, "room    ") {}
+      m_socket(ip, port, "room    ") {
+    m_passcode = passcode;
```
Comment (stblr) [inline payload/sp/cs/RoomClient.cc:101]: Not possible to do it like the other ones?

File: payload/game/ui/DirectConnectionPage.cc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    m_replacement = PageId::OnlineTop;
+    startReplace(Anim::Next, 0.0f);
+}
+
+void DirectConnectionPage::onDigitButtonFront([[maybe_unused]] PushButton *button,
+        [[maybe_unused]] u32 localPlayerId) {
+    m_editBox.insert(button->m_index);
+}
+
+void DirectConnectionPage::onOkButtonFront(PushButton *button, [[maybe_unused]] u32 localPlayerId) {
```
Comment (stblr) [inline payload/game/ui/DirectConnectionPage.cc:84]: What happens if the code is not full yet? The vanilla game has some logic to enable/disable this button in 8061e348, 8061e77c, 8061e270.

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/game/ui/DirectConnectionPage.cc (hunk 3 above)
Comment (vabold) [reply payload/game/ui/DirectConnectionPage.cc:94]: Resolved in https://github.com/stblr/mkw-sp/pull/455/commits/4273b1c6dc32710ad2c956da21467d9307bbb11f.

File: payload/game/ui/DirectConnectionPage.cc (hunk 1 above)
Comment (vabold) [reply payload/game/ui/DirectConnectionPage.cc:3]: https://github.com/stblr/mkw-sp/pull/455/commits/4273b1c6dc32710ad2c956da21467d9307bbb11f

File: payload/game/ui/DirectConnectionPage.cc [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    m_editBox.insert(button->m_index);
+    if (m_editBox.isFull()) {
+        m_okButton.setPlayerFlags(1);
+        m_okButton.select(0);
+    }
+}
+
+void DirectConnectionPage::onOkButtonFront(PushButton *button, [[maybe_unused]] u32 localPlayerId) {
+    u64 directCode = m_editBox.getNumber();
+    if (directCode - 0x3ed5142afa4755f > directCode || directCode > 0xbed51428fa4755e) {
```
Comment (stblr) [inline payload/game/ui/DirectConnectionPage.cc:96]: The first comparison can be simplified.

File: payload/game/ui/Button.hh [hunk 5]
```diff
@@ -54,6 +54,7 @@ public:
     void setPlayerFlags(u32 playerFlags);
     void setPointerOnly(bool pointerOnly);
     f32 getDelay() const;
+    bool isSelected();
```
Comment (stblr) [inline payload/game/ui/Button.hh:57]: `const`

Comment (vabold) [reply payload/game/ui/Button.hh]: Resolved in https://github.com/stblr/mkw-sp/pull/455/commits/acd8f5bb5a66d7a5167e19e1ce13fae8c944e744.

File: payload/game/ui/DirectConnectionPage.cc (hunk 4 above)
Comment (vabold) [reply payload/game/ui/DirectConnectionPage.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/455/commits/dfd6b2b324246c66f76b4b232d1c7c3e112b8cd7.

**Review by stblr — APPROVED** [review]

**Pull request merged**